src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2265,107 @@ private void EndEntry()
+
+                        IEnumerable<IEdmProperty> selectedProperties = resourceState.SelectedProperties.GetSelectedProperties(edmStructuredType);
+
+                        // All dynamic properties: null value restoration for properties that are not present.
+                        foreach (string name in resourceState.SelectedProperties.GetSelectedDynamicProperties(edmStructuredType))
+                        {
+                            if (resource.Properties.Any(p => p.Name.Equals(name, StringComparison.Ordinal))
+                                || resourceState.NavigationPropertiesRead.Contains(name))
+                            {
+                                continue;
+                            }
+
+                            RestoreNullODataProperty(name, resourceState);
+                        }
+
+                        foreach (IEdmProperty currentProperty in selectedProperties)
+                        {
+                            Debug.Assert(currentProperty.Type != null, "currentProperty.Type != null");
+
+                            // Skip declared properties that are not null-able types.
+                            // Skip navigation properties (no navigation links need to be restored as null.)
+                            // Skip properties that have been read (primitive or structural).
+                            if (!currentProperty.Type.IsNullable
+                                || currentProperty.PropertyKind == EdmPropertyKind.Navigation
+                                || resource.Properties.Any(p => p.Name.Equals(currentProperty.Name, StringComparison.Ordinal))
+                                || resourceState.NavigationPropertiesRead.Contains(currentProperty.Name))
+                            {
+                                continue;
+                            }
+
+                            RestoreNullODataProperty(currentProperty.Name, resourceState);
+                        }
+
+                        // Restore null Expanded entities, which are specified as "isExpandedNavigationProperty" the child nodes.
+                        foreach (string name in resourceState.SelectedProperties.NamesOfSelectedExpandedEntities)
+                        {
+                            if (!resource.IsExpandedEntityProcessed(name))
+                            {
+                                RestoreNullODataProperty(name, resourceState);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Restore the null property value in the resource.
+        /// </summary>
+        /// <param name="nullPropertyName">Name of the property to restore the null value.</param>
+        /// <param name="resourceState">The resource state.</param>
+        private static void RestoreNullODataProperty(string nullPropertyName, IODataJsonLightReaderResourceState resourceState)

```

Comment:
>RestoreNullODataProperty [](start = 28, length = 24)

Seems inefficient to attempt a cast and do a concat for each individual property being restored.  Could we instead pass a collection of properties and add them in a single concatenation?

Reply:
The original ReadOnlyEnumerable only supports adding an item to the underneath list. I just expand the API for adding a list of items, so that we can add list of items once for all from here.
Now we pass a collection and add all of the items at once.

---
In reply to: [230799899](https://github.com/OData/odata.net/pull/1106#discussion_r230799899) [](ancestors = 230799899)